### ISSUE TITLE
Events eviction/DLQ

### ISSUE THREAD

**Issue by colinsurprenant** [elastic/logstash#5283]
In the context of upcoming internal queuing persistence, we will introduce pipeline-wide reliability to make sure that queued events will not be removed until acknowledged as fully processed by the final pipeline stage. This will ensure at-least-once delivery semantic in particular in the case of an unsupervised stop or a crash. 

In that context, events flowing in the pipeline will require to **always** be acknowledged in the normal runtime. The only situation where events are not acknowledged is upon an unsupervised stop or a crash. This requirement to always acknowledge events introduces the need to evict events upon processing errors or catched/uncatched exceptions.

These should be the possible outcomes of event processing:

```
  1. the process is successful, the event/batch is ACK'ed
  2. the process is blocked, ACK cannot happens until the process is unblocked
  3. the process crashes with an uncatched exception, the event/batch is ACK'ed and evicted (DLQ)
  4. the process catches an exception:
    4.1. if there's a retry mechanism, the event/batch is retried
      4.1.1. the retry is successful, the event/batch is ACK'ed 
      4.1.2. the retry fails (max retry?), the event/batch is ACK'ed and evicted (DLQ)
    4.2. there's no retry mechanism, the event/batch is ACK'ed and evicted (DLQ)
```

We need to introduce a proper/holistic logstash-wide events eviction mechanism.

**Comment by jordansissel** [comment]
+1 on the proposal. In terms of implementation, if ack-and-evict isn't atomic, we'll want to evict to the DLQ and ACK *after* the DLQ write succeeds to avoid data loss on machine failures.

**Comment by acchen97** [comment]
> 3. the process crashes with an uncatched exception, the event/batch is ACK'ed and evicted (DLQ)

The ack/evict option makes sense for pipeline crashes, but should we also entertain the scenario of continued processing (from non-empty queue) post restart?  It would need to restart with the same pre-crash config, but seems like a notable use case to consider.

I agree this queue eviction stuff can and should seemingly converge with other error routing cases you noted here, sharing similar "DLQ" type storage infra.  This may be a completely separate discussion, but re: destinations for eviction/error routing, there's a few options and we could potentially architect with an easy workflow for reprocessing this data back through LS.  I feel the initial focus could be just writing to local file or ES (or simply discard if desired), thoughts?  Btw, ingest node did implement some [failure processing](https://www.elastic.co/guide/en/elasticsearch/reference/master/handling-failure-in-pipelines.html) to route error events into another ES index.

**Comment by colinsurprenant** [comment]
@acchen97 
> continued processing (from non-empty queue) post restart

Do you mean post restart after *normal* termination? If so, yes, restarting after a shutdown with non-empty queues, should just resume processing normally and process queued events.

I think DLQ could take many shapes and I agree that whatever we choose will require a comprehensive way for reprocessing this evicted data. One possible solution would be to provide the configuration of a DLQ (global? per stage? per plugin?) using our plugins and config syntax just like an output. For example, we could configure a file output with json codec as a DLQ. If we do this, we'll need to figure how to deal with evictions from the DLQ "stage" to avoid a DLQ-DLQ problem.

**Comment by acchen97** [comment]
@colinsurprenant 

> Do you mean post restart after normal termination? 

What defines normal termination? A graceful shutdown shouldn't leave queues unflushed, so is this 2x ctrl+c or --allow-unsafe-shutdown?  Regardless of termination type, I feel there's value in having both options of continued processing or ack/evict when restarting the pipeline.

> One possible solution would be to provide the configuration of a DLQ (global? per stage? per plugin?) using our plugins and config syntax just like an output.

Per plugin instance error handling granularity would be ideal, i.e. S3 output errors could simply just drop the event, while ES output errors should send to DLQ.

**Comment by colinsurprenant** [comment]
@acchen97 

> What defines normal termination? A graceful shutdown shouldn't leave queues unflushed

I am not sure about this. In the situation where there's back pressure because of slow output and persisted queues are growing, a shutdown could very well just complete the current output operation and terminate gracefully, leaving the non-empty persisted queues in place and upon restart just resume where it left off?

**Comment by colinsurprenant** [comment]
@acchen97 to recap, let me know if this maps correctly to your uses-cases: these are the possible termination scenarios related to persistence:
- events backlog accumulated in the PQ due to slower outputs (backpressure)
  - a normal shutdown would terminate the current **in-flight events** (not those in the PQ), finalizing the ack'ing or dlq'ing of these events and terminate logstash. upon restart, logstash would pickup where it left in the PQ,  processing the events backlog.
  - a crash would leave the inflight events as unack'ed in the PQ and upon restart these would be first  replayed and logstash would then pickup where it left in the PQ,  processing the events backlog.

- no PQ backlog accumulation due to slower inputs (no backpressure)
  - a normal shutdown would terminate the current in-flight events, finalizing the ack'ing or dlq'ing of these events and terminate logstash. Chances are that the DLQ would be empty or almost empty and upon restart, logstash would pickup where it left in the PQ.
  - a crash would leave the inflight events as unack'ed in the PQ and upon restart these would be replayed and logstash would pickup where it left in the almost empty PQ.

**Comment by acchen97** [comment]
@colinsurprenant thanks for the recap. I'm wondering if PQs are coupled with a certain config. For instance, what happens when a LS instance has an existing non-empty PQ and is restarted with a *different* config? I'm thinking it should either directly flush the existing PQ to DLQ  before resuming or have some type of confirmation prompt prior to flush.

Also, if we haven't defined a DLQ file structure and re-ingest workflow, I'd be happy to chat about it. As the event can be in different states based on pipeline stage at time of dead lettering, it could be tricky.

**Comment by suyograo** [comment]
As a first implementation, or a poor man's DLQ we can log the full, unacked events in JSON format to just evict them. Additionally, using the new log4j2 infrastructure, one can create a special appender for this package. For example: `logger.persistent.queues.unacked_events` goes to a separate file. This way, log4j2 takes care of rotation and keeping things at bay. 

Also, it may be nice to introduce a new setting `path.unacked_events` which by default points to `path.data/unacked_events`. We can reference this setting in the log4j.properties. With this, the DLQ becomes fully configurable and has the ability to rotate so we don't fill the disk. 

As a user, I can then take this file and simply use file input with json codec to pipe it to a separate LS instance and process it further.

**Comment by colinsurprenant** [comment]
@suyograo followup issue #5948

**Comment by guyboertje** [comment]
Firstly, I find the word eviction troublesome. For the acronym ETL, the point of the Load step is to place the Event in the designated Destination(s) and if this cannot be done then it needs to be put in a designated Alternate Destination.

Eviction is troublesome for me because the Event is not really evicted, it is acked and drops from the PQ either way - a failed Event is simply put in (one of) the Alternate Destination(s).

I even find the idea of coding for **one and only one** DLQ troublesome.

1. **The user** may regard some transforms as critical to deciding whether an Event should be further transformed and loaded
2. **The user** may regard some transforms as non-critical and may wish for the Event to be routed to an Alternative destination for manual enhancement but maybe not.
3. **The user** may regard some destinations as critical for a successful Load
4. **The user** may regard some destinations as non-critical for a successful Load

When deemed unsuccessful the Event should be tagged and considered failed.

We now have the Batch object. It should be used to segregate the "failed" events from the active ones (it already segregates the cancelled ones).

We have previously discussed a universal error handling inside the pipeline, filters or their base class or filter_delegate so that batches are always acknowledged. We should definitely include a config option that allows a user to specify the criticality of a stage and this option should guide how we tag the event and whether it should be sent to an alternative destination or not.

**Comment by jordansissel** [comment]
+1 "evict" isn't the right term. We are delivering the data to an alternate
location (a dead-letter queue) upon failure to deliver normally.

**Comment by colinsurprenant** [comment]
I am not sure I understand the semantic argumentation on "eviction". I think the idea behind is very clear: at **some point** (a) an event is failing normal processing/delivery and (b) must be sent to an alternate route. 

"Eviction" was never meant as "dropping" since obviously we are actually talking about solving the event dropping problem by introducing Dead Letter Queuing. 

Now, how we deal with (a) to decide when an event is failed and (b) where that event should go is what we need to come up with a coherent DLQ story but clearly this will have to be done in steps because trying to tackle all the possibilities as listed above in @guyboertje example is way ahead of what we have today which is basically nothing: failed events are dropped. I believe we can move forward by providing reasonable and practical options and at the same time see our users are using this and improve.
